data/data.go

```diff
@@ -116,34 +127,105 @@ func extractInt(v map[string]interface{}, field string, default_val int) int {
 	return int(float)
 }
 
+func (r *QueryResponder) handleScrollId(users []string, sorts []elastic.SortInfo, scrollId string, scroll string) {
+	source, err := r.buildSearchSource(users, sorts)
+	if err != nil {
+		return
+	}
+
+	res, err := r.es.Scroll().SearchSource(source).ScrollId(scrollId).Scroll(scroll).Do(r.ctx)
+	if err != nil {
+		r.writer.WriteHeader(http.StatusBadRequest)
+		r.logAndOutputErr(err)
+		return
+	}
+
+	r.outputSearchResults(res)
+}
+
+func (r *QueryResponder) buildSearchSource(users []string, sorts []elastic.SortInfo) (*elastic.SearchSource, error) {
+	usersJson, err := json.Marshal(users)
+	if err != nil {
+		r.writer.WriteHeader(http.StatusInternalServerError)
+		r.logAndOutputErr(err)
+		return nil, err
+	}
+	permFieldScript := `
+		String perm = null;
+		for (up in params._source.userPermissions) {
+			for (user in ` + string(usersJson) + `) {
+				if (up.user == user && perm != 'own' && !(up.permission == 'read' && perm == 'write')) {
```

Comment:
Can't really comment on the Go code, but seems ok!  Was only curious about the purpose of this permission check.

Reply:
Well, this is the thing that calculates the provided user's permission level on each result. `perm` is the current calculated highest permission level. So, if it's already `own`, nothing can make it higher. If it's at the `write` level, then it's fine to replace it as long as it's either `write` (changing nothing) or `own`, which works out to "not `read`". Thus: for each entry in the document, for each user/group that the user belongs to, if the permission entry in the doc is for that user, set the calculated permission to that permission as long as it's the same or a higher level of permission (i.e., if the calculated one isn't already ownership and it's not downgrading a write to a read).

But it's definitely a bit convoluted, lol.

Reply:
I think I'll add a comment explaining the script. That seems like the thing to do here.

Reply:
Got it - thanks for the explanation!